src/tracing/metric_macros.h

```diff
@@ -0,0 +1,177 @@
+
+// Utility to always return the 3rd argument from macro parameters
+#define __GET_3RD_ARGUMENT(_a1, _a2, _a3, ...) _a3
+
+// Utility macro to select the VerifyOrExit macro with the correct signature based on the invoked macro
+#define __SELECT_MACRO_FOR_VERIFY_ACTION(...) __GET_3RD_ARGUMENT(__VA_ARGS__, __VERIFY_ACTION_2ARGS, __VERIFY_ACTION_1ARGS, )
+
+// Wrapper to capture all arguments and invoke the real wrapper for VerifyOrExit's third argument
+#define __LOG_METRIC_FOR_VERIFY_ACTION(...) __SELECT_MACRO_FOR_VERIFY_ACTION(__VA_ARGS__)(__VA_ARGS__)
+
+// Utility macro that takes the action wraps it to emit the metric in a Verify macro, when a metric key is specified
+#define __VERIFY_ACTION_2ARGS(anAction, metricKey) MATTER_LOG_METRIC(metricKey, anAction)
+
+// Macro that takes the action in a Verify macro when a metric key is not specified
+#define __VERIFY_ACTION_1ARGS(anAction) anAction
+
+// Utility macro used by VerifyOrExit macro to handle an optional third metric key argument and emit an event, if specified
+#define LOG_METRIC_FOR_VERIFY_OR_EXIT_ACTION(anAction, ...) __LOG_METRIC_FOR_VERIFY_ACTION(anAction, ##__VA_ARGS__)
+
+// Utility macro to select the SuccessOrExit macro with the correct signature based on the invoked macro
+#define __SELECT_MACRO_FOR_SUCCESS_OR_EXIT(...) __GET_3RD_ARGUMENT(__VA_ARGS__, __SUCCESS_OR_EXIT_2ARGS, __SUCCESS_OR_EXIT_1ARGS, )
+
+// Wrapper to capture all arguments and invoke the real wrapper for SuccessOrExit's second argument
+#define __LOG_METRIC_FOR_SUCCESS_OR_EXIT(...) __SELECT_MACRO_FOR_SUCCESS_OR_EXIT(__VA_ARGS__)(__VA_ARGS__)
+
+// Utility macro that takes the status and wraps it to emit the metric in a SuccessOrExit macro, when a metric key is specified
+#define __SUCCESS_OR_EXIT_2ARGS(aStatus, metricKey)                                                                                \
+    chip::Tracing::utils::logMetricIfError(aStatus, chip::Tracing::kMetric##metricKey)
+
+// Utility macro that just evaluates the status when no metric key is specified
+#define __SUCCESS_OR_EXIT_1ARGS(aStatus) ::chip::ChipError::IsSuccess((aStatus))
+
+// Utility macro used by SuccessOrExit macro to handle an optional second metric key argument and emit an event, if specified
+#define LOG_METRIC_FOR_SUCCESS_OR_EXIT(aStatus, ...) __LOG_METRIC_FOR_SUCCESS_OR_EXIT(aStatus, ##__VA_ARGS__)
+
+////////////////////// Metric LOGGING
+
+// Utility to always return the 4th argument from macro parameters
+#define __GET_4TH_ARG(_a1, _a2, _a3, _a4, ...) _a4
+
+// Utility macro to select the macro with the correct signature based on the invoked MATTER_LOG_METRIC_XYZ macro
+#define __SELECT_MACRO_FOR_EVENT_METRIC(...)                                                                                       \
+    __GET_4TH_ARG(__VA_ARGS__, __MATTER_LOG_METRIC_3ARGS, __MATTER_LOG_METRIC_2ARGS, __MATTER_LOG_METRIC_1ARGS, )
+
+// Wrapper to capture all arguments and invoke the real wrapper for logging metrics
```

Comment:
We should better document all this black art macro

Reply:
Reworked a large portion. The macro magic still exists for the LOG metric to support overloading